modules/oe_content_event/src/Plugin/InternalLinkSourceFilter/EventPeriodFilter.php

```diff
@@ -0,0 +1,176 @@
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->timeBasedCacheTagGenerator = $time_based_cache_tag_generator;
+    $this->time = $time;
+    $this->entityTypeManager = $entity_type_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('oe_time_caching.time_based_cache_tag_generator'),
+      $container->get('datetime.time'),
+      $container->get('entity_type.manager')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function apply(QueryInterface $query, array $context, RefinableCacheableDependencyInterface $cacheability): void {
+    $now = new DrupalDateTime();
+    $current_time = $this->time->getCurrentTime();
+    $now->setTimestamp($current_time);
+    $now->setTimezone(new \DateTimeZone(DateTimeItemInterface::STORAGE_TIMEZONE));
+    switch ($this->getConfiguration()['period']) {
+      case self::PAST:
+        $query->condition('oe_event_dates.end_value', $now->format(DateTimeItemInterface::DATETIME_STORAGE_FORMAT), "<");
+        $query->sort('oe_event_dates.end_value', 'DESC');
+        // Time based cache tags need to be invalidated when a new event
+        // ends so find out which event is going to end next and extract
+        // its end date.
+        $results = $this->entityTypeManager->getStorage('node')->getQuery()
+          ->condition('oe_event_dates.end_value', $now->format(DateTimeItemInterface::DATETIME_STORAGE_FORMAT), ">")
+          ->sort('oe_event_dates.end_value', 'ASC')
+          ->execute();
+        $next_ending_event = $this->entityTypeManager->getStorage('node')->load(reset($results));
+        $next_ending_event_datetime = new DrupalDateTime($next_ending_event->oe_event_dates->end_value, new \DateTimeZone(DateTimeItemInterface::STORAGE_TIMEZONE));
+        $cacheability->addCacheTags($this->timeBasedCacheTagGenerator->generateTags($next_ending_event_datetime->getPhpDateTime()));
+        break;
+
+      case self::UPCOMING:
+        $query->condition('oe_event_dates.value', $now->format(DateTimeItemInterface::DATETIME_STORAGE_FORMAT), ">");
+        $query->sort('oe_event_dates.value', 'ASC');
+        // Time based cache tags need to be invalidated when the next
+        // upcoming event starts so execute the query and extract the
+        // starting date of the first event in the list.
+        $results = $query->execute();
+        $next_starting_event = $this->entityTypeManager->getStorage('node')->load(reset($results));
+        $next_starting_event_datetime = new DrupalDateTime($next_starting_event->oe_event_dates->value, new \DateTimeZone(DateTimeItemInterface::STORAGE_TIMEZONE));
```

Comment:
Let's not use the query we are altering to fetch results, I'm not sure if running execute() doesn't change any internal state for that object.

I'd create a separate method and use the Entity Query service for that, it might look as a duplication but it's more conceptually correct, I'd say.